### PR TITLE
Improve project_panel diagnostic icon knockout colors

### LINKED ISSUE #20572: Error indicators in project panel should use different border for selected files
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

In #18182 error indicators were added. For the selected file, the border used for the error icon is of  an incorrect color

Screenshot 2024-11-13 at 10 26 35

### Environment

Zed: v0.162.0 (Zed Nightly 55cd99cdc4d22bf82c0a59d8fa76f4ab9b85370c)
OS: macOS 15.1.0
Memory: 8 GiB
Architecture: aarch64

### PR DESCRIPTION
**Pull request by nilskch**
Closes #20572

Release Notes:

- N/A

cc @danilo-leal @WeetHet

### CODE REVIEW

File: crates/project_panel/src/project_panel.rs [hunk 1]
```diff
@@ -2762,11 +2783,12 @@ impl ProjectPanel {
                 }
             }))
             .cursor_pointer()
+            .bg(bg_color)
             .child(
                 ListItem::new(entry_id.to_proto() as usize)
                     .indent_level(depth)
                     .indent_step_size(px(settings.indent_size))
-                    .selected(is_marked || is_active)
+                    .selectable(false)
```
Comment (nilskch) [inline crates/project_panel/src/project_panel.rs:2791]: This was a tricky one. We had a `.hover` style adjustments in the div that contained the ListItem AND ListItem sets the background color as well if the ListItem is selectable. `selectable` does not change the behaviour of click events or anything else, it only controls the coloring.

File: crates/project_panel/src/project_panel.rs [hunk 2]
```diff
@@ -101,6 +101,7 @@ pub struct ProjectPanel {
     // We keep track of the mouse down state on entries so we don't flash the UI
     // in case a user clicks to open a file.
     mouse_down: bool,
+    hovered_entries: HashSet<ProjectEntryId>,
```
Comment (nilskch) [inline crates/project_panel/src/project_panel.rs:104]: Do the listener callbacks run potentially on a different thread and this HashSet should be threadsafe?

**Comment by nilskch** [comment]
This is how it looks like:

https://github.com/user-attachments/assets/1ca677b3-318d-4d30-9eaa-ab2bd169f47f


There is one thing that is missing. We do dynamic style adjustments for the drag_over color here:

https://github.com/zed-industries/zed/blob/516f7b364213b2bbc1a658295df96443aa885ca0/crates/project_panel/src/project_panel.rs#L2693

I would really like to maintain a state for the `drag_over_entries` just like we did with the `hovered_entries`, but unfortunately there is no `on_drag_over` function that i can pass a listener to like we did with `on_hover`. 

I think we should add this function (either in this PR or separately) to the `div` and make sure that the drag over color matches as well.

File: crates/project_panel/src/project_panel.rs [hunk 3]
```diff
@@ -2625,6 +2630,14 @@ impl ProjectPanel {
             marked_selections: selections,
         };
 
+        let (bg_color, border_color) = match (is_hovered, is_marked || is_active, self.mouse_down) {
+            (true, _, true) => (item_colors.marked_active, item_colors.hover),
+            (true, false, false) => (item_colors.hover, item_colors.hover),
+            (true, true, false) => (item_colors.hover, item_colors.marked_active),
+            (false, true, _) => (item_colors.marked_active, item_colors.marked_active),
+            _ => (item_colors.default, item_colors.default),
+        };
```
Comment (nilskch) [inline crates/project_panel/src/project_panel.rs:2639]: The idea is to keep a state for all the events and not do dynamic style adjustments with `.hover(|style| ...`, so we can pass the same color to the background and to the knockout icons. I think this logic replicates exactly the previous behaviour.

**Review by danilo-leal — APPROVED** [review]
This looks great to me—thank you again, @nilskch! I took a stab at intercepting the hover status last week and knew it would require something similar to this. I'd like an additional pair of eyes here before merging, though—if either @osiewicz or @mrnugget could check out the code!

**Comment by osiewicz** [comment]
I wonder if we can achieve the same result without explicitly tracking hover state of an entry.

File: crates/project_panel/src/project_panel.rs (hunk 2 above)
Comment (osiewicz) [reply crates/project_panel/src/project_panel.rs:104]: Yeah, listeners run on foreground IIRC, so we should be good here. Besides, it wouldn't compile if it wasn't thread-safe 😄

**Comment by bennetbo** [comment]
Paired on this with @danilo-leal. Although I'm not a big fan of tracking the hovered entries solely for the purpose of the UI, we did not find a straight-forward way to achieve this without tracking the state explicitly.
Would be great if anyone can come up with an approach without tracking this state explicitly, but let's get this in for now!

**Pull request merged**